src/data_cell/flatten_datacell.h

```diff
@@ -0,0 +1,194 @@
+        return this->factoryComputerImpl(query);
+    }
+
+    float
+    ComputePairVectors(uint64_t id1, uint64_t id2) override;
+
+    void
+    Train(const float* data, uint64_t count) override;
+
+    void
+    InsertVector(const float* vector, uint64_t idx) override;
+
+    void
+    BatchInsertVector(const float* vectors, uint64_t count, uint64_t* idx) override;
+
+    void
+    SetMaxCapacity(uint64_t capacity) override {
+        this->max_capacity_ = std::max(capacity, this->total_count_);  // TODO add warning
+    }
+
+    [[nodiscard]] uint64_t
+    TotalCount() const override {
+        return this->total_count_;
+    }
+
+    [[nodiscard]] const uint8_t*
+    GetCodesById(uint64_t id) const override;
+
+    inline void
+    SetQuantizer(std::shared_ptr<Quantizer<QuantTmpl>> quantizer) {
+        this->quantizer_ = quantizer;
+        this->code_size_ = quantizer_->GetCodeSize();
+    }
+
+    inline void
+    SetIO(std::shared_ptr<BasicIO<IOTmpl>> io) {
+        this->io_ = io;
+    }
+
+public:
+    uint64_t total_count_{0};
+
+    uint64_t max_capacity_{1000000};
+
+    uint64_t code_size_{0};
+
+    std::shared_ptr<Quantizer<QuantTmpl>> quantizer_{nullptr};
+
+    std::shared_ptr<BasicIO<IOTmpl>> io_{nullptr};
+
+private:
+    inline void
+    queryImpl(float* resultDists, const float* queryVector, const uint64_t* idx, uint64_t idCount);
+
+    inline void
+    queryImpl(float* resultDists,
+              std::shared_ptr<Computer<QuantTmpl>> computer,
+              const uint64_t* idx,
+              uint64_t idCount);
+
+    inline std::shared_ptr<ComputerInterface>
+    factoryComputerImpl(const float* query) {
+        auto computer = this->quantizer_->FactoryComputer();
+        computer->SetQuery(query);
+        return computer;
+    }
+};
+
+template <typename QuantTmpl, typename IOTmpl>
+FlattenDataCell<QuantTmpl, IOTmpl>::FlattenDataCell(const std::string& initializeJson) {
+    // TODO
+}
+
+template <typename QuantTmpl, typename IOTmpl>
+void
+FlattenDataCell<QuantTmpl, IOTmpl>::Train(const float* data, uint64_t count) {
+    if (this->quantizer_) {
+        this->quantizer_->Train(data, count);
+    }
+}
+
+template <typename QuantTmpl, typename IOTmpl>
+void
+FlattenDataCell<QuantTmpl, IOTmpl>::InsertVector(const float* vector, uint64_t idx) {
+    if (idx == std::numeric_limits<uint64_t>::max()) {
+        idx = total_count_;
+        ++total_count_;
+    }
+    auto* codes = new uint8_t[code_size_];
+    quantizer_->EncodeOne(vector, codes);
+    io_->Write(codes, code_size_, idx * code_size_);
+    delete[] codes;
+}
+
+template <typename QuantTmpl, typename IOTmpl>
+void
+FlattenDataCell<QuantTmpl, IOTmpl>::BatchInsertVector(const float* vectors,
+                                                      uint64_t count,
+                                                      uint64_t* idx) {
+    if (idx == nullptr) {
+        auto* codes = new uint8_t[code_size_ * count];
```

Comment:
where is the place to process the alloc failure?

Reply:
we will use allocator to alloc and free in future